Shell Recap
============
The shell is a program where users can type commands. With the shell, it’s possible to invoke complicated programs like climate modeling software or simple commands that create an empty directory with only one line of code. The most popular Unix shell is Bash. Bash is the default shell on most modern implementations of Unix and in most packages that provide Unix-like tools for Windows.

The grammar of a shell allows you to combine existing tools into powerful pipelines and handle large volumes of data automatically. Sequences of commands can be written into a script, improving the reproducibility of workflows.

The command line is often the easiest way to interact with remote machines and supercomputers. Familiarity with the shell is nearly essential to run a variety of specialized tools and resources including high-performance computing systems. As clusters and cloud computing systems become more popular for scientific data crunching, being able to interact with the shell is becoming a necessary skill. We can build on the command-line skills covered here to tackle a wide range of scientific questions and computational challenges.

Let's start whit this recap!

**DISCLAIMER:** since we are in the Jupyter environment, all the commands or the cells that exploit bash functionalities will be preceded by the "**!**" or by "**%%bash**" this is necessary to make the Jupyter able to correctly interpreted the source code as bash code. 
In order to run the same command in a real bash, you have to remove the "**!**" and "**%%bash**"

## The Basics

### The variables
The variables in the shell have a different behavior in comparison to other languages, their declaration is simple

In [ ]:
%%bash
NAME="Shell Scripting is Fun!"

the command used to print a variable is:

In [1]:
%%bash
NAME="Shell Scripting is Fun!"
echo $NAME

Shell Scripting is Fun!


Running the same command with only the variable name will produce the printing of the variable name, but not the content of the var.

In [2]:
%%bash
NAME="Shell Scripting is Fun!"
echo NAME

NAME


### The print command

The ```echo``` command is a command used to print at screen a variable or whichever data you need to show.
As we have seen in the previous cells, with echo you can also concatenate variable/commands and strings. Let's see.

In [3]:
%%bash
APPLES=3
echo "Pippo has $APPLES apples"


Pippo has 3 apples


In [4]:
%%bash

echo "I'm in the working folder $(pwd)"

I'm in the working folder /home/scampese/Repository/ManagementAndAnalysisOfPhysicsDatasetsB/lecture_1


The option ">" or ">>" respectively represent the "insert" and "append" of data to a file, the echo command can be used to put whatever you want to print to a file.

### The first command:

The first command that you have to remember is the ```pwd``` that stands for *print working directory*.
This command can be used to see/understand in which folder we are working.
Let's see an example

In [5]:
! pwd

/home/scampese/Repository/ManagementAndAnalysisOfPhysicsDatasetsB/lecture_1


### How to see what's in the folder:

The 'ls' command is used to see what the folder contains.

In [6]:
%%bash 
ls

bash_shell_recap.ipynb
exercise
lecture_1.zip
virtualbox-guide.pdf


The ls command has several parameters that allow you to see what you're looking for. One of those parameters if ```-F```. 
This parameter can be used with other to understand if each entry on the list is a file or a folder. To run this cell on your pc, you have to create a temporary test folder in the working folder


In [7]:
%%bash
ls -F

bash_shell_recap.ipynb
exercise/
lecture_1.zip
virtualbox-guide.pdf


In general every available parameter of a command can be seen by running the command with the ```--help``` parameter.

In [8]:
%%bash

ls --help

Uso: ls [OPZIONE]... [FILE]...
List information about the FILEs (the current directory by default).
Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.

Mandatory arguments to long options are mandatory for short options too.
  -a, --all                  non nasconde le voci che iniziano con .
  -A, --almost-all           non elenca le voci . e ..
      --author               con -l stampa l'autore di ogni file
  -b, --escape               stampa escape in stile C per i caratteri non grafici
      --block-size=SIZE      scale sizes by SIZE before printing them; e.g.,
                               '--block-size=M' prints sizes in units of
                               1,048,576 bytes; see SIZE format below
  -B, --ignore-backups       do not list implied entries ending with ~
  -c                         with -lt: sort by, and show, ctime (time of last
                               modification of file status information);
                               with -l:

### Let's create a new folder

The ```mkdir``` command is used to create a new folder starting from where you are. If you want you can also pass to the command the whole path you want to put the new folder in.
Let's try!

In [9]:
%%bash
mkdir newTemporaryFolder
ls -F

bash_shell_recap.ipynb
exercise/
lecture_1.zip
newTemporaryFolder/
virtualbox-guide.pdf


To create a new folder inside the created one the commands becomes:

In [10]:
%%bash 

mkdir newTemporaryFolder/newTemporaryFolderInside
ls ./newTemporaryFolder -F

newTemporaryFolderInside/


### How to create a simple empty file?

The ```touch``` command is a simple command that allows you to create an empty file with the name and the extension that you want. Here's a simple example

In [11]:
%%bash

touch tempFile.txt
ls -F

bash_shell_recap.ipynb
exercise/
lecture_1.zip
newTemporaryFolder/
tempFile.txt
virtualbox-guide.pdf


### How to go in/out from a folder?

The ```cd``` command has two functions. Getting in or getting out of a folder. To enter a folder the command becomes ```cd folderName```. If you want to enter in a sub-folder, you can do this just by concatenating the folders that contain it: ```cd folderName/folder2Name/..```. 

If you want to exit a folder the command becomes ```cd ..```. Similar to the entering, if you want to get out of several folders, you have to concatenate the dots as ```cd ../..```.

If you want to come back to your $HOME directory, from wherever you are, this command gives you the shortcut ```cd ```.

Let's try to cd in a folder and then get out.

In [12]:
%%bash

cd newTemporaryFolder
echo "I'm in:" $(pwd)
cd ..
echo "Now I'm in:" $(pwd)

I'm in: /home/scampese/Repository/ManagementAndAnalysisOfPhysicsDatasetsB/lecture_1/newTemporaryFolder
Now I'm in: /home/scampese/Repository/ManagementAndAnalysisOfPhysicsDatasetsB/lecture_1


### How to delete a file or a folder

The ```rm``` command is a command needed when you want to delete a file. If you want to delete both the folder an its content, the command becomes ```rm -r```.
Let's see a couple of examples

In [13]:
%%bash

echo "Before deleting the file:"
ls -F
echo "******************************************"

echo "After deleting the file:"
rm tempFile.txt
ls -F
echo "******************************************"

echo "After deleting the folder:"
rm -r newTemporaryFolder
ls -F

Before deleting the file:
bash_shell_recap.ipynb
exercise/
lecture_1.zip
newTemporaryFolder/
tempFile.txt
virtualbox-guide.pdf
******************************************
After deleting the file:
bash_shell_recap.ipynb
exercise/
lecture_1.zip
newTemporaryFolder/
virtualbox-guide.pdf
******************************************
After deleting the folder:
bash_shell_recap.ipynb
exercise/
lecture_1.zip
virtualbox-guide.pdf


### How to Copy or Move a file or a folder

To copy a file or a folder you need to use ```cp```, while the command ```mv``` lets you move a file or a folder.
Mv command can also be used to rename a file: "*Moving a file into another name*"



In [14]:
%%bash 
mkdir startingPath
mkdir destinationPath
touch startingPath/file.txt

mkdir pippo

cp startingPath/file.txt destinationPath/file.txt


In [15]:
! ls destinationPath

file.txt


In [16]:
%%bash 

rm  startingPath/file.txt
mv destinationPath/file.txt startingPath/file.txt

In [17]:
! ls destinationPath

In [18]:
! ls startingPath

file.txt


But ```mv``` can be used even to rename...

In [19]:
%%bash

mv pippo pluto

In [20]:
!ls 

bash_shell_recap.ipynb	exercise       pluto	     virtualbox-guide.pdf
destinationPath		lecture_1.zip  startingPath


### How to see the content of a file

For this purpose, there are several options. The most common integrated command in the Unix Shell are:
+ cat
+ tail
+ head

The ```cat``` command can be used to get all the content of a file or a set of files and print them to the screen regardless of the file format (binary file included). This command is very useful but it is not appreciable when you have to deal with thousands and million-row files, which is quite common in the Big Data era.

The ``` head``` command allows you to get the first n rows of a file where n is a parameter of the command. This file can be useful when you need a fast check on a file structure or when you need to look for something in the first rows.

The ```tail``` command allows you to get the last n rows of a file where n is a parameter of the command. This command is particularly helpful when you have to check a log of a program (i.e.: you want to check just the last iteration/execution of a program and not the whole log file). Moreover, ```tail``` allows you to "*listen*" to a file that is being written and print in real-time what is going to be put on that file. This is quite useful when you have a log of your cluster and you want to check what is happening to the computation.


In [21]:
%%bash 

touch tempFile.txt
echo "Hello to MAPD 19/20" > tempFile.txt
cat tempFile.txt


Hello to MAPD 19/20


In [22]:
%%bash 

echo "The first lecture is about bash" >> tempFile.txt
cat tempFile.txt

Hello to MAPD 19/20
The first lecture is about bash


In [23]:
%%bash
head -n 1 tempFile.txt

Hello to MAPD 19/20


In [24]:
%%bash 
tail -n 1 tempFile.txt

The first lecture is about bash


### Search a file/directory

the ```find``` command is used to search a file or a folder. It has a huge set of different options to make really complex searches. Let's see a simple example:


In [25]:
%%bash 
find . -name tempFile.txt

./tempFile.txt


The aforementioned command is used to search, starting from our working directory, the tempFile.txt. In the next cell there's a command that can be used to search all the files with the txt extension that are present in our working directory. To change the searched directory we need to change the . with the path fo the folder you want to search in.

In [26]:
%%bash 
find . -name '*.txt' ## the qoutes around the argument are not necessay out of Jupyter

./startingPath/file.txt
./exercise/exercise_file_3.txt
./exercise/exercise_file_2.txt
./exercise/exercise_file_4.txt
./exercise/exercise_file_1.txt
./exercise/exercise_file_5.txt
./exercise/haiku.txt
./tempFile.txt


In [27]:
%%bash 

mkdir newTemporaryFolder
touch ./newTemporaryFolder/anotherTestFile.txt
find ./newTemporaryFolder -name '*.txt' ## the qoutes around the argument are not necessay out of Jupyter

rm -r newTemporaryFolder

./newTemporaryFolder/anotherTestFile.txt


## Search in a file

The ```grep``` command can be used to search something in a file. 
Let's see an example:

In [28]:
%%bash
grep MAPD tempFile.txt

Hello to MAPD 19/20


grep allows to extract all the line that contains the word that we want to search. This command can also work with regular expressions.

In [29]:
%%bash
grep -E "[0-9]+" tempFile.txt

Hello to MAPD 19/20


## Counting words or lines of a file:

the ```wc``` command allows you to count the number of words in a file:

In [30]:
%%bash
wc -w tempFile.txt

10 tempFile.txt


or the number of line

In [31]:
%%bash
wc -l tempFile.txt

2 tempFile.txt


## Sorting
the command ```sort``` can be used to sort the elements in a file or in a list of elements:

In [32]:
%%bash

echo "5" > orderable.txt
echo "3" >> orderable.txt
echo "6" >> orderable.txt
echo "2" >> orderable.txt

sort orderable.txt

2
3
5
6


## Complex but useful commands

### How can we build a pipeline of commands?

the answer is with char ```|```. 
This is a special char that can be used to create a pipeline of commands that must be executed in order. 
For example when searching in a file:

In [33]:
%%bash 
cat tempFile.txt | grep MAPD

Hello to MAPD 19/20


Or it can be used to order a set of files by the number of words inside (ascending):

In [34]:
%%bash 
echo "one two three" > file1.txt
echo "one two three four" > file2.txt
echo "one" > file3.txt


wc -w file* | sort -n 

 1 file3.txt
 3 file1.txt
 4 file2.txt
 8 totale


## Getting the list of the processes in the system

```ps``` is a command that allows to list all the processes that are running in the system.


In [35]:
! ps aux


USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.0 225424  9228 ?        Ss   13:33   0:02 /sbin/init spla
root         2  0.0  0.0      0     0 ?        S    13:33   0:00 [kthreadd]
root         3  0.0  0.0      0     0 ?        I<   13:33   0:00 [rcu_gp]
root         4  0.0  0.0      0     0 ?        I<   13:33   0:00 [rcu_par_gp]
root         6  0.0  0.0      0     0 ?        I<   13:33   0:00 [kworker/0:0H-k
root         9  0.0  0.0      0     0 ?        I<   13:33   0:00 [mm_percpu_wq]
root        10  0.0  0.0      0     0 ?        S    13:33   0:00 [ksoftirqd/0]
root        11  0.0  0.0      0     0 ?        I    13:33   0:02 [rcu_sched]
root        12  0.0  0.0      0     0 ?        S    13:33   0:00 [migration/0]
root        13  0.0  0.0      0     0 ?        S    13:33   0:00 [idle_inject/0]
root        14  0.0  0.0      0     0 ?        S    13:33   0:00 [cpuhp/0]
root        15  0.0  0.0      0     0 ?        S    13:33   0:00 [

Where:
+ a = show processes for all users
+ u = display the process user/owner
+ x = also show processes not attached to a terminal

The PID of the process can be used to kill a stuck process (i.e. when you run a python script that is going to be stuck)


The ```ps``` command can be used with the ```grep``` command through the ```|``` character, in order to search for a particular process.

In [36]:
%%bash 
ps aux | grep python

root       907  0.0  0.0 170544 17232 ?        Ssl  13:33   0:00 /usr/bin/python3 /usr/bin/networkd-dispatcher --run-startup-triggers
root      1196  0.0  0.0 187248 19884 ?        Ssl  13:33   0:00 /usr/bin/python3 /usr/share/unattended-upgrades/unattended-upgrade-shutdown --wait-for-signal
scampese  3497  0.0  0.2 651956 65252 pts/0    Sl+  13:36   0:03 /usr/bin/python3 /usr/local/bin/jupyter-notebook
scampese  3602  0.0  0.1 617380 47964 ?        Ssl  13:36   0:00 /usr/bin/python3 -m ipykernel_launcher -f /home/scampese/.local/share/jupyter/runtime/kernel-d7b4b4f2-bd62-4360-ad0b-7a7663936864.json
scampese  3635  0.0  0.1 617380 48368 ?        Ssl  13:37   0:00 /usr/bin/python3 -m ipykernel_launcher -f /home/scampese/.local/share/jupyter/runtime/kernel-f005c673-0bb4-4906-a3eb-5aac4bda7bb4.json
root      3714  0.0  0.0 167912 17204 ?        Sl   13:37   0:00 /usr/bin/python3 /usr/share/apt-xapian-index/update-apt-xapian-index-dbus
scampese  5630  0.0  0.1 626120 49620 ?        Ssl  13

sudo kill -9 907

## Figuring out the IP address
The ```ifconfig``` command can be used to retrieve the Ip address. 
This command is not a standard command of the Unix system, and in some cases it must be installed. To install it on ubuntu you need to run this command:
    ```sudo apt-get install net-tools```

In [37]:
%%bash

ifconfig

enp8s0f1: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether 80:fa:5b:73:35:77  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Loopback locale)
        RX packets 6853  bytes 5484280 (5.4 MB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 6853  bytes 5484280 (5.4 MB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlp7s0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.101  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 fe80::c5dd:34c7:c8bd:c911  prefixlen 64  scopeid 0x20<link>
        ether 4c:1d:96:6d:28:06  txqueuelen 1000  (Ethernet)
        RX p

the output of the command reports all the "network card" and the ip addresses.

### Other useful command in the near future

Other commands that may be useful in the near future are the ```ssh``` command that allows you to remotely connect to a server, cluster or HPC, and ```scp``` that is a command necessary to copy local-to-remote or remote-to-local files and resources.

[https://www.digitalocean.com/community/tutorials/ssh-essentials-working-with-ssh-servers-clients-and-keys](https://www.digitalocean.com/community/tutorials/ssh-essentials-working-with-ssh-servers-clients-and-keys)

[https://haydenjames.io/linux-securely-copy-files-using-scp/](https://haydenjames.io/linux-securely-copy-files-using-scp/)

Those two command exploit the same ssh protocol to communicate with a remote device.